File: 207-course-schedule/course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        indegree = [0] * numCourses
        adj = [[] for _ in range(numCourses)]
        for x,y in prerequisites:
            indegree[x]+=1
            adj[y].append(x)
        
        q = deque( i for i in range(numCourses) if indegree[i] == 0)
        ans = []
        
        while q:
            cur = q.popleft()
            ans.append(cur)
            for ch in adj[cur]:
                indegree[ch]-=1
                if indegree[ch] == 0:
                    q.append(ch)

        return True if len(ans) == numCourses else False

File: 207-course-schedule/course-schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<int>in(numCourses);
        vector<vector<int>>adj(numCourses);
        for(auto pre : prerequisites){
            adj[pre[1]].push_back(pre[0]);
            in[pre[0]]++;
        }
        queue<int>q;
        int cnt = 0;
        for(int i = 0 ; i <numCourses ;i++){
            if(!in[i])q.push(i),cnt++;
        }
        while(q.size()){
            int node = q.front();q.pop();
            for(auto ch : adj[node]){
                in[ch]--;
                if(!in[ch]){
                    cnt++;
                    q.push(ch);
                }
                
            }
        }
        return cnt == numCourses;
    }
};